### ISSUE TITLE
Docker containers on diffrent VM Network -> No failover triggered even a new etcd leader elected

### ISSUE THREAD

**Issue by mlyahmed** [zalando/patroni#1347]
Hi,

We are actually trying to find out an effective pattern to insure a HA on our PG plateform. We would like to use Patroni to insure PG HA, we want to run it as docker container. For some special reasons related to our context, we, actually, can't use docker-compose neither swarm. Thus, the expected MVP plateform will
is a set of VMs running native docker images. We are heading to Kubernetes on the long term.

Now we are facing an issue using patroni. It does not behave as expected in case of network split. we expect a failover but it does not even a new etcd leader is elected.

We tested that in two orchestrations cases. It does not result the expectation and behaves diffrently.

The tests are set up on AWS EC2 and we are ready to share that with you: 3 DB VM holding Patroni PG and Patroni ETCD + 2 VMs holding Patroni HAProxy + an App

Orchesration 1 : https://patroni-docker-test.s3.eu-west-3.amazonaws.com/patroni-docker-orchesration-1.png
	Here the patroni/PG container is peered with the local and remote etcd instances.

Orchesration 2 : https://patroni-docker-test.s3.eu-west-3.amazonaws.com/patroni-docker-orchesration-2.png
	Here the patroni/PG container is peered with the local etcd node instance. 

Test : When we completly  isolate the primary, by just keeping only the ssh port open for the inbound and outbound traffic, a new etcd leader is elected but no failover is trigered on the Patroni/PG side :

We can see that the ETCD leader is demoted :
`2019-12-26 12:26:25.530426 I | raft: 90f282fef93e9fb8 is starting a new election at term 66
2019-12-26 12:26:25.530986 I | raft: 90f282fef93e9fb8 became candidate at term 67
2019-12-26 12:26:25.531262 I | raft: 90f282fef93e9fb8 received vote from 90f282fef93e9fb8 at term 67
2019-12-26 12:26:25.531577 I | raft: 90f282fef93e9fb8 [logterm: 1, index: 3] sent vote request to d7e50f5affc975a5 at term 67
2019-12-26 12:26:25.531856 I | raft: 90f282fef93e9fb8 [logterm: 1, index: 3] sent vote request to f5ebccccf101eb83 at term 67
2019-12-26 12:26:25.536013 I | raft: 90f282fef93e9fb8 received vote rejection from f5ebccccf101eb83 at term 67
2019-12-26 12:26:25.536393 I | raft: 90f282fef93e9fb8 [quorum:2] has received 1 votes and 1 vote rejections
2019-12-26 12:26:25.537191 I | raft: 90f282fef93e9fb8 received vote rejection from d7e50f5affc975a5 at term 67
2019-12-26 12:26:25.537487 I | raft: 90f282fef93e9fb8 [quorum:2] has received 1 votes and 2 vote rejections
2019-12-26 12:26:25.537773 I | raft: 90f282fef93e9fb8 became follower at term 67
2019-12-26 12:26:26.579785 I | raft: 90f282fef93e9fb8 [term: 67] received a MsgVote message with higher term from d7e50f5affc975a5 [term: 68]
2019-12-26 12:26:26.580419 I | raft: 90f282fef93e9fb8 became follower at term 68
2019-12-26 12:26:26.580751 I | raft: 90f282fef93e9fb8 [logterm: 1, index: 3, vote: 0] voted for d7e50f5affc975a5 [logterm: 9, index: 149] at term 68
2019-12-26 12:26:26.583942 I | raft: raft.node: 90f282fef93e9fb8 elected leader d7e50f5affc975a5 at term 68
2019-12-26 12:26:26.587417 N | etcdserver: set the initial cluster version to 2.2
2019-12-26 12:26:26.588104 N | etcdserver: updated the cluster version from 2.2 to 2.3
2019-12-26 12:26:26.589621 I | etcdserver: published {Name:etcd1 ClientURLs:[http://172.31.10.105:2379]} to cluster 6daf44502dc964fc
2019-12-26 12:26:27.439336 W | rafthttp: the connection to peer d7e50f5affc975a5 is unhealthy
2019-12-26 12:26:27.440390 W | rafthttp: the connection to peer f5ebccccf101eb83 is unhealthy
2019-12-26 12:26:32.382824 I | rafthttp: the connection with d7e50f5affc975a5 became active
2019-12-26 12:26:32.383859 I | rafthttp: the connection with f5ebccccf101eb83 became active`

And a new one is elected
`2019-12-26 12:25:52.584789 W | rafthttp: the connection to peer 90f282fef93e9fb8 is unhealthy
2019-12-26 12:26:22.585250 W | rafthttp: the connection to peer 90f282fef93e9fb8 is unhealthy
2019-12-26 12:26:25.011560 I | rafthttp: the connection with 90f282fef93e9fb8 became active
2019-12-26 12:26:25.080394 N | etcdserver: updated the cluster version from 2.2 to 2.3
2019-12-26 12:26:25.533782 I | raft: d7e50f5affc975a5 [term: 9] received a MsgVote message with higher term from 90f282fef93e9fb8 [term: 67]
2019-12-26 12:26:25.534024 I | raft: d7e50f5affc975a5 became follower at term 67
2019-12-26 12:26:25.534064 I | raft: d7e50f5affc975a5 [logterm: 9, index: 149, vote: 0] rejected vote from 90f282fef93e9fb8 [logterm: 1, index: 3] at term 67
2019-12-26 12:26:25.534089 I | raft: raft.node: d7e50f5affc975a5 lost leader f5ebccccf101eb83 at term 67
2019-12-26 12:26:26.576884 I | raft: d7e50f5affc975a5 is starting a new election at term 67
2019-12-26 12:26:26.577126 I | raft: d7e50f5affc975a5 became candidate at term 68
2019-12-26 12:26:26.577171 I | raft: d7e50f5affc975a5 received vote from d7e50f5affc975a5 at term 68
2019-12-26 12:26:26.577193 I | raft: d7e50f5affc975a5 [logterm: 9, index: 149] sent vote request to f5ebccccf101eb83 at term 68
2019-12-26 12:26:26.577213 I | raft: d7e50f5affc975a5 [logterm: 9, index: 149] sent vote request to 90f282fef93e9fb8 at term 68
2019-12-26 12:26:26.581675 I | raft: d7e50f5affc975a5 received vote from f5ebccccf101eb83 at term 68
2019-12-26 12:26:26.582021 I | raft: d7e50f5affc975a5 [quorum:2] has received 2 votes and 0 vote rejections
2019-12-26 12:26:26.582135 I | raft: d7e50f5affc975a5 became leader at term 68
2019-12-26 12:26:26.582482 I | raft: raft.node: d7e50f5affc975a5 elected leader d7e50f5affc975a5 at term 68
2019-12-26 12:30:40.278165 E | rafthttp: **failed to read 90f282fef93e9fb8 on stream Message (read tcp 172.18.0.2:51054->172.31.10.105:2380: i/o timeout)**
2019-12-26 12:30:40.278399 I | rafthttp: the connection with 90f282fef93e9fb8 became inactive
2019-12-26 12:30:40.380342 I | rafthttp: the connection with 90f282fef93e9fb8 became active`

But on postgres side, the primary is kept and no failover is triggered.

Please note that we had to change the Patroni Docker entrypoint.sh file to mount patroni and etcd based on the EC2 external IP, we call it PATRONI_EXTERNAL_IP, instead of DOCKER_IP since the containers are run on diffrent networks outside Docker's. Please checkout the file https://patroni-docker-test.s3.eu-west-3.amazonaws.com/entrypoint.sh

Do I miss something ?

Happy end-of-year festivities !
Ahmed

**Comment by CyberDem0n** [comment]
Do you realize that etcd leader and patroni leader are two absolutely different things?
Have you ever looked into patroni logs?

**Comment by mlyahmed** [comment]
I do know that they are diffrent things. 

What I can't explain is that since the whole VM is isolated from the others, etcd patroni/elects a new leader, patroni/pg is cut off from others exept local etcd, that became folower, and no failover is trigerred.

on patroni/pg logs nothing changed, the other nodes keep tracing that they are folowers and the lock is hold by patroni1 and patroni1 keep saying he holds the lock and then he's the leader.

**Comment by mlyahmed** [comment]
Another detail is that once the VM is isolated on patroni folowers trace an error and keep running as standby 
`2019-12-26 13:08:41.918 UTC [3056] FATAL:  could not connect to the primary server: could not connect to server: Connection timed out
                Is the server running on host "172.31.10.105" and accepting
                TCP/IP connections on port 5432?
2019-12-26 13:08:44,864 INFO: Lock owner: patroni1; I am patroni2
2019-12-26 13:08:44,864 INFO: does not have lock
2019-12-26 13:08:44,872 INFO: no action.  i am a secondary and i am following a leader
2019-12-26 13:08:54,868 INFO: Lock owner: patroni1; I am patroni2
`

on the primary at a time it terminates walsender but keeps running as leader:
`
2019-12-26 12:31:35.202 UTC [92] LOG:  terminating walsender process due to replication timeout
2019-12-26 12:31:39.677 UTC [112] LOG:  terminating walsender process due to replication timeout
2019-12-26 12:31:44,854 INFO: Lock owner: patroni1; I am patroni1
2019-12-26 12:31:44,867 INFO: no action.  i am the leader with the lock

`

**Comment by CyberDem0n** [comment]
The leader lock doesn't expire instantly. It takes up to ttl seconds before the failover is triggered. 
And again, you didn't showed logs from patroni nodes and didn't showed the configuration.

**Comment by CyberDem0n** [comment]
JFYI, if Patroni lost connection to one etcd node, it retries with another one.
Hightly likely you "isolation" isn't good enough.

**Comment by mlyahmed** [comment]
I noticed that since all patroni instances are peered with all etcd instances. But that behavior I noticed it on the follower nodes not on the isolated node. I even tried some curl to check the isolation and it seems effective.

Ok let me try again and get  back to you.

Here are the scripts I use to launch instances.
These scripts are just for testing.
Here is how the patroni/etcd are launched on each VM https://patroni-docker-test.s3.eu-west-3.amazonaws.com/etcd
Here is how the patroni/pg are launched on each VM https://patroni-docker-test.s3.eu-west-3.amazonaws.com/postgres
And the HaProxy https://patroni-docker-test.s3.eu-west-3.amazonaws.com/haproxy
The image mlyahmed/patroni:1.2.0 is now public

I retry again the scenarios and get back to you with full logs

**Comment by mlyahmed** [comment]
Regarding the TTL I didn't find that in the documentation. Surely I miss read it. How can I adjust it please ?
In the first try I recall leaving the nodes some minutes to see what will happen and nothing've changed. Not more then 5 minutes.

**Comment by CyberDem0n** [comment]
Magic doesn't exists.
If the primary is able to reach one of the etcd nodes and update the leader lock it will continue to run as primary.

The default value of ttl is 30 seconds. In case is the primary is really isolated (can't write into etcd cluster), it will demote itself within 20 seconds.

**Comment by CyberDem0n** [comment]
The fact that it can't reach etcd is not possible to miss, there will be a lot of exceptions and stack traces in logs.

**Comment by mlyahmed** [comment]
We know magic does not exist in IT.
I am trying to reproduce it and testing the isolation. Get back to you
thanks

**Comment by mlyahmed** [comment]
Here are the logs :
On VM 1 :
	patroni/etcd : https://patroni-docker-test.s3.eu-west-3.amazonaws.com/atcd1.log
	patroni/pg : https://patroni-docker-test.s3.eu-west-3.amazonaws.com/patroni1.log
On VM 2 :
	patronig/etcd : https://patroni-docker-test.s3.eu-west-3.amazonaws.com/etcd2.log
	patronig/pg : https://patroni-docker-test.s3.eu-west-3.amazonaws.com/patroni2.log
On VM 3 :
	patronig/etcd : https://patroni-docker-test.s3.eu-west-3.amazonaws.com/etcd3.log
	patronig/pg : https://patroni-docker-test.s3.eu-west-3.amazonaws.com/patroni3.log
	
regarding the isolation. Here is the AWS SG applied to the VM1. It is the only SG applied to the VM :
	inbound : https://patroni-docker-test.s3.eu-west-3.amazonaws.com/aws-sg-sshonly-inbound.PNG
	outbound : https://patroni-docker-test.s3.eu-west-3.amazonaws.com/aws-sg-sshonly-outbound.PNG

**Comment by mlyahmed** [comment]
A healthcheck from the VM1 gave https://patroni-docker-test.s3.eu-west-3.amazonaws.com/healthcheck.PNG

So now we got VM1 where there is etcd as folower, patroni as leader and the complete VM is isolated. Is it normal ?
Thanks

**Comment by CyberDem0n** [comment]
At the beginning of every log there is a line: `INFO: Selected new etcd server http://172.31.39.72:2379`.
In case if Patroni lost access to the etcd-cluster (the specific node) it will switch to another node and there should be another similar line in the logs.
Since patroni logs contain no other similar lines, all patroni nodes are **still successfully connected** to `http://172.31.39.72:2379`. I am pretty sure you can verify it with tcpdump.

**Comment by mlyahmed** [comment]
Here is the end result. HaProxy does not see any master https://patroni-docker-test.s3.eu-west-3.amazonaws.com/haproxy-healthcheck.PNG so the application is down.

Let me please check that connection to http://172.31.39.72:2379

get back to you

**Comment by CyberDem0n** [comment]
etcd nodes communicating between each other via port 2380, and etcd logs show that isolation sort of working.
HAProxy communicates with Patroni via REST API, and apparently it can't reach it (this is expected).
Patroni communicates with etcd via port 2379, and for some reason connections are still working.

That could only mean that isolation doesn't properly work.

**Comment by mlyahmed** [comment]
Ok. Got a snapshot https://patroni-docker-test.s3.eu-west-3.amazonaws.com/eth0.172_31_39_72.pcap
seems they still some successful commucation here is a shark http snapshot https://patroni-docker-test.s3.eu-west-3.amazonaws.com/shark-snapshot.PNG

How come this isolation doesn't work properly since I cut all inbound and outbound rules ?

**Comment by mlyahmed** [comment]
> etcd nodes communicating between each other via port 2380, and etcd logs show that isolation sort of working.
> HAProxy communicates with Patroni via REST API, and apparently it can't reach it (this is expected).
> Patroni communicates with etcd via port 2379, and for some reason connections are still working.
> 
> That could only mean that isolation doesn't properly work.

Now here is the situation.

**on VM 1**. etcd lost leader and a new one is elected. Even with an AWS SG allowing ssh/22 only, the etcd1 successfully keep communication with new leader :

`2019-12-26 14:50:01.377608 I | raft: replace the unstable entries from index 3186
2019-12-26 14:50:01.377665 I | raft: raft.node: 90f282fef93e9fb8 elected leader d7e50f5affc975a5 at term 44
2019-12-26 14:51:07.178561 E | rafthttp: failed to write d7e50f5affc975a5 on stream Message (write tcp 172.18.0.2:2380->172.31.39.72:44198: i/o timeout)
2019-12-26 14:51:07.178763 I | rafthttp: the connection with d7e50f5affc975a5 became inactive
2019-12-26 14:51:07.265284 I | rafthttp: the connection with d7e50f5affc975a5 became active
2019-12-26 15:00:20.517338 E | rafthttp: failed to heartbeat f5ebccccf101eb83 on stream Message (write tcp 172.18.0.2:2380->172.31.17.42:47090: i/o timeout)
2019-12-26 15:00:20.517530 I | rafthttp: the connection with f5ebccccf101eb83 became inactive
2019-12-26 15:05:13.841887 E | rafthttp: failed to heartbeat d7e50f5affc975a5 on stream MsgApp v2 (write tcp 172.18.0.2:2380->172.31.39.72:44196: i/o timeout)
2019-12-26 15:05:13.842114 I | rafthttp: the connection with d7e50f5affc975a5 became inactive
2019-12-26 15:05:13.864989 I | rafthttp: the connection with d7e50f5affc975a5 became active
2019-12-26 15:33:40.977306 I | etcdserver: start to snapshot (applied: 10001, lastsnap: 0)
2019-12-26 15:33:40.993838 I | etcdserver: saved snapshot at index 10001
2019-12-26 15:33:40.994287 I | etcdserver: compacted raft log at 5001
2019-12-26 15:52:36.184627 I | rafthttp: the connection with f5ebccccf101eb83 became active`

According to tcpdump on the VM1, strangely the old leader, etcd1, keep syncing with the new leader, etcd3, by putting queries as it his the effective leader and etcd3 respond OK. https://patroni-docker-test.s3.eu-west-3.amazonaws.com/etcd1-put-himself-as-leader-and-etcd3-accepts-it.PNG
More then that, it requests its status and etcd3 responds it that it still the leader https://patroni-docker-test.s3.eu-west-3.amazonaws.com/etcd3-responds-etcd1-as-this-last-is-master.PNG


**On the other side on VM3**. etcd3 keeps logged as it is the new leader and at same time complaining it is not able to communicate with etcd1 :
`2019-12-26 14:50:01.365694 I | raft: d7e50f5affc975a5 is starting a new election at term 43
2019-12-26 14:50:01.365920 I | raft: d7e50f5affc975a5 became candidate at term 44
2019-12-26 14:50:01.365977 I | raft: d7e50f5affc975a5 received vote from d7e50f5affc975a5 at term 44
2019-12-26 14:50:01.366062 I | raft: d7e50f5affc975a5 [logterm: 43, index: 3185] sent vote request to 90f282fef93e9fb8 at term 44
2019-12-26 14:50:01.366133 I | raft: d7e50f5affc975a5 [logterm: 43, index: 3185] sent vote request to f5ebccccf101eb83 at term 44
2019-12-26 14:50:01.366185 I | raft: raft.node: d7e50f5affc975a5 lost leader 90f282fef93e9fb8 at term 44
2019-12-26 14:50:01.376079 I | raft: d7e50f5affc975a5 received vote from f5ebccccf101eb83 at term 44
2019-12-26 14:50:01.376273 I | raft: d7e50f5affc975a5 [quorum:2] has received 2 votes and 0 vote rejections
2019-12-26 14:50:01.376349 I | raft: d7e50f5affc975a5 became leader at term 44
2019-12-26 14:50:01.376442 I | raft: raft.node: d7e50f5affc975a5 elected leader d7e50f5affc975a5 at term 44
2019-12-26 14:50:02.587540 W | etcdserver: failed to reach the peerURL(http://172.31.10.105:2380) of member 90f282fef93e9fb8 (Get http://172.31.10.105:2380/version: dial tcp 172.31.10.105:2380: i/o timeout)
2019-12-26 14:50:02.587736 W | etcdserver: cannot get the version of member 90f282fef93e9fb8 (Get http://172.31.10.105:2380/version: dial tcp 172.31.10.105:2380: i/o timeout)
2019-12-26 14:50:05.343148 E | rafthttp: failed to read 90f282fef93e9fb8 on stream Message (read tcp 172.18.0.2:44198->172.31.10.105:2380: i/o timeout)
2019-12-26 14:50:05.343322 I | rafthttp: the connection with 90f282fef93e9fb8 became inactive`


The same **on VM2**. etcd2 is following etcd3 and keep complaining it is not able to communicate with etcd1 :
`2019-12-26 14:32:51.776016 E | rafthttp: failed to dial d7e50f5affc975a5 on stream MsgApp v2 (dial tcp 172.31.39.72:2380: getsockopt: connection refused)
2019-12-26 14:32:51.776243 I | rafthttp: the connection with d7e50f5affc975a5 became inactive
2019-12-26 14:33:01.484340 I | rafthttp: the connection with d7e50f5affc975a5 became active
2019-12-26 14:50:01.372471 I | raft: f5ebccccf101eb83 [term: 43] received a MsgVote message with higher term from d7e50f5affc975a5 [term: 44]
2019-12-26 14:50:01.372681 I | raft: f5ebccccf101eb83 became follower at term 44
2019-12-26 14:50:01.372745 I | raft: f5ebccccf101eb83 [logterm: 43, index: 3185, vote: 0] voted for d7e50f5affc975a5 [logterm: 43, index: 3185] at term 44
2019-12-26 14:50:01.372828 I | raft: raft.node: f5ebccccf101eb83 lost leader 90f282fef93e9fb8 at term 44
2019-12-26 14:50:01.377569 I | raft: raft.node: f5ebccccf101eb83 elected leader d7e50f5affc975a5 at term 44
2019-12-26 14:50:05.343276 E | rafthttp: failed to read 90f282fef93e9fb8 on stream Message (read tcp 172.18.0.2:47090->172.31.10.105:2380: i/o timeout)
2019-12-26 14:50:05.343302 I | rafthttp: the connection with 90f282fef93e9fb8 became inactive
2019-12-26 15:06:05.326799 W | rafthttp: the connection to peer 90f282fef93e9fb8 is unhealthy
2019-12-26 15:06:35.327152 W | rafthttp: the connection to peer 90f282fef93e9fb8 is unhealthy`

Besides that the isolation does not work as expected, is it a normal situation where :

- **etcd1 demoted himself but keep puting etcd3 as he is the leader and this former keeps saying OK.**
- **At the same time, etcd3 elected as leader and can't even connect to etcd1.**

?

**Comment by mlyahmed** [comment]
And finally it is a situation where by the end the whole stack is broken since the application is down because it can't update any data and the cluster can't detect and heal himself.

**Comment by mlyahmed** [comment]
it seems that etcd1 keeps also communication with etcd2 https://patroni-docker-test.s3.eu-west-3.amazonaws.com/shark-snapshot-etcd2.PNG

Here is a snapshot : https://patroni-docker-test.s3.eu-west-3.amazonaws.com/eth0.172_31_17_42.pcap

**Comment by CyberDem0n** [comment]
All Patroni nodes can write to etcd cluster, but other ports are isolated. It is totally not a fault of Patroni that you failed to do the proper isolation.

**Comment by CyberDem0n** [comment]
Patroni promises are very simple: if it can write to etcd - it is allowed to run postgres as primary, so it does.

**Comment by mlyahmed** [comment]
I stopped etcd1 and keep seeing exchanges between VM1 and etcd3 : https://patroni-docker-test.s3.eu-west-3.amazonaws.com/eth0.172_31_39_72_after_stopping_etcd1.pcap

Now when I stoppped patroni1, a failover is triggered and no traces shown on tcpdump.

VM 2 :
	etcd2 : https://patroni-docker-test.s3.eu-west-3.amazonaws.com/etcd2-failover.log
	patroni2: https://patroni-docker-test.s3.eu-west-3.amazonaws.com/patroni2-failover.log
	
VM 3 :
	etcd 3 : https://patroni-docker-test.s3.eu-west-3.amazonaws.com/etcd3-failover.log
	patroni 3 : https://patroni-docker-test.s3.eu-west-3.amazonaws.com/patroni3-failover.log
	
Having said that, I admit that the isolation seems not complete and will dig further to check why. But this failure that is for you "human" could happen at any time automatically due to a network or system failure. I suggest to keep the eye on the no-human case and take a position if it is accepted as siituation or not. And if not how the whole must behave.

If you accept it, thing that is not insuring for me, we can stop here.

Just remind the final situation : HAProxy can't find any master and the application is down and at the same time there is a PG primary locked out there.

Regards

**Comment by CyberDem0n** [comment]
If there is a network partitioning it is very unlikely that the partitioned node is still accessible via some ports and not accessible via other, it is either all or nothing. If it happened anyway, I have a bad news for you, you got much bigger problem than unreachable primary.

Regarding HAProxy - it simply can't access the node running as primary (I already mentioned it twice, this is the third time). Since HAProxy can't access the primary (because of you firewalled it), the application can't reach it either.

**Comment by mlyahmed** [comment]
> All Patroni nodes can write to etcd cluster, but other ports are isolated. It is totally not a fault of Patroni that you failed to do the proper isolation.

Please check my previous comment ! ectd1 stopped and communication is kept. patroni1 stopped and failover is triggered and patroni3 is elected as primary next to ectd3 that was already leader.

**Comment by mlyahmed** [comment]
> If there is a network partitioning it is very unlikely that the partitioned node is still accessible via some ports and not accessible via other, it is either all or nothing. If it happened anyway, I have a bad news for you, you got much bigger problem than unreachable primary.
> 
> Regarding HAProxy - it simply can't access the node running as primary (I already mentioned it twice, this is the third time). Since HAProxy can't access the primary (because of you firewalled it), the application can't reach it either.

Regarding HAPRoxy and the application I reminded you the situation that could happen at any time which we want to avoid. You don't need to explain le why the application is down or why HAProxy does not find the master.

The all or nothing happens when network split but in case of system/firewall failure this is not garanteed.

**Comment by CyberDem0n** [comment]
> Please check my previous comment ! ectd1 stopped and communication is kept. 

Please also read my comments. Patroni can connect to **ANY** etcd node. In case if one node fails it transparently switches to another one.

> patroni3 is elected as primary next to ectd3 that was already leader.

Once again, patroni leader and etcd leader are two absolutely different things. It is not required to run patroni and etcd on the same nodes. Think about it.

>The all or nothing happens when network split but in case of system/firewall failure this is not garanteed.

This is ridiculous. If somebody is messing up with the firewall you are in bigger trouble than just master not being accessible.
With the same result someone might just close access to HAProxy and applications will not be able to reach the database. It is not a Patroni task to manage your firewall rules.
